packages/utilities/src/utilities/scrollIntoView/scrollIntoView.js

```diff
@@ -0,0 +1,176 @@
+/**
+ * Copyright IBM Corp. 2016, 2020
+ *
+ * This source code is licensed under the Apache-2.0 license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+import { breakpoints } from '@carbon/layout';
+
+/**
+ * Amount of columns used for calculation.
+ *
+ * @private
+ */
+
+const _colSpan = 3;
+
+/**
+ * The inner viewport calculation for the root margins.
+ *
+ * @private
+ */
+
+function _getViewportMargin() {
+  return (
+    '-' +
+    (
+      (document.documentElement.clientHeight * _colSpan) /
+      breakpoints.max.columns
+    ).toString() +
+    'px 0px'
+  );
+}
+
+/**
+ * Intersection Observer options
+ *
+ * @private
+ */
+
+const _options = {
+  rootMargin: _getViewportMargin(),
+  threshold: 0,
+};
+
+/**
+ * Iteration boolean for continuous play option.
+ *
+ * @private
+ */
+
+let _iterations;
+
+/**
+ * Intersection Observer that watches outer viewport.
+ *
+ * @private
+ */
+
+let _rootObserver;
+
+/**
+ * Intersection observer that watches the inner viewport.
+ *
+ * @private
+ */
+
+let _innerObserver;
+
+/**
+ * Resize observer to trigger rootMargin recalculations
+ *
+ * @private
+ */
+let _resizeObserver;
+
+/**
+ * Utility handles fade transition for selected elements.
+ *
+ * @example
+ * import '@carbon/ibmdotcom-styles/scss/internal/scroll-into-view/_scroll-into-view.scss';
+ * import { scrollIntoView } from '@carbon/ibmdotcom-utilities';
+ *
+ * As an example, the function can be called to target all instances of the
+ * elements in a list:
+ *
+ * const list = ['.bx--content-block', 'bx--content-group'];
+ *
+ * For default values of 400ms and 'one and done' play:
+ * scrollIntoView(list);
+ *
+ * With 'continuous play' option:
+ * scrollIntoView(list, { iterations: true })
+ *
+ * For custom delay time, set within targeted class in the application's CSS code as such:
+ *
+ * .bx--content-block {
+ *   --#{$dds-prefix}--scroll-into-view-delay: 250ms;
+ * }
+ *
+ * @param {*} elementList of all the elements to be targeted
+ * @param {boolean} iterations to define whether its continuous or not
+ */
+
+const scrollIntoView = (elementList, { iterations = false } = {}) => {
+  _iterations = iterations;
+
+  window.addEventListener(
+    'load',
+    () => {
+      _rootObserver = new IntersectionObserver(handleExit);
+      _innerObserver = new IntersectionObserver(handleEntrance, _options);
+      _resizeObserver = new ResizeObserver(handleResize);
+
+      elementList.forEach(selector => {
+        let elements = document.querySelectorAll(selector);
+        elements.forEach(e => {
+          _rootObserver.observe(e);
+          _innerObserver.observe(e);
+        });
+      });
+      _resizeObserver.observe(document.documentElement);
+    },
+    false
+  );
+};
```

Comment:
I now see one thing that slipped my mind... My apologies that I didn't notice thing earlier. We need to detach event listeners, `IntersectionObserver` and `ResizeObserver` when the target DOM nodes are removed. Such use case often happens with "routers" (e.g. React routers) when navigating to different "route" swaps DOM and thus the event listeners, etc. will be orphaned (and causes memory leaks, JavaScript errors, etc).

A quick way to solve this is adding something like below:

```javascript
return {
  release() {
    _resizeObserver.disconnect();
    _innerObserver.disconnect();
    _rootObserver.disconnect();
    window.removeEventListener('load', handleOnLoad);
  },
};
```

Another way (that I'd prefer given scroll into view thing now effectively has a lifecycle) is making it a component rather than a function. We can do it by one of the following:

* Create React component as well as Web Component
* Use Carbon core vanilla API if we still want to make framework neutral: https://github.com/carbon-design-system/carbon/tree/v10.25.0/packages/components/src/globals/js/mixins#component-lifecycle-create-componentjs

CC @jeffchew

Reply:
For performance purposes a component wrapper might make more sense for this. Perhaps we can build web component wrapper first, and would we be able to create the react wrapper of that or a pure react component would make more sense @asudoh @IgnacioBecerra ?

Reply:
@jeffchew Web Component and a React wrapper on top of it makes sense to me - @IgnacioBecerra what do you think?

Reply:
@jeffchew @asudoh Hm, just so I understand, would creating a Web Component for this would be something like:
```
// fade-component.ts
class DDSFadeComponent {
   // scroll-into-view code here + the disconnect thing
}

// dotcom-shell.stories.ts
let elementsToFade = "dds-content-block, dds-content-item, etc..."
let _iterations = true;
.
.
.
return html`
  <dds-fade-component elements="${elementstoFade}" iterations="${_iterations}"/>
  <dds-table-of-contents>
     // etc etc 
`
```

After typing it out like this I think I understand it better, but let me know if my understanding is wrong.

Reply:
Great question @IgnacioBecerra! We have two choices:

1. Wrap each element (or groups of elements) for fade in/out effect with `<dds-fadeinout>`
    * Pro
        * Idiomatic wrt what set of elements need fade in/out effect
        * Strict control of the lifecycle
    * Con:
        * Extra HTML elements
2. Make each scrolling element `<dds-fadeinout>`
    * Pro
        * Less extra HTML elements
    * Con
        * Still need to specify target elements (via CSS selector)
        * Still have possibility of memory leaks (if the target elements have different lifecycle from `<dds-fadeinput>`)

Basically, if we tend to (or can) add fade in/out effect to "region"s or components rather than some indivisual (specific) elements, 1. works the best (e.g. are they only heading/copy elements that need fade in/out effect in content block?). Otherwise 2. works better.